File: sdk/memory-core/typescript/src/cos.ts
/**
 * Memory file reader — direct read of memory pipeline artifacts (persona.md,
 * scene_blocks/*.md) from object storage with STS credential management.
 *
 * Usage:
 *   const reader = createMemoryFileReader({ endpoint, apiKey, serviceId });
 *   const content = await reader.read("scene_blocks/cooking-recipes.md");
 *
 * Under the hood:
 *   1. Fetches STS temporary credentials from the platform (POST /v2/cos/secret)
 *   2. Caches credentials until they expire (auto-refresh with 2-min buffer)
 *   3. Signs a COS V5 GET request with the STS credentials
 *   4. Returns the file content as a string
 *
 * Storage backend (currently COS) is an implementation detail — the public
 * API is intentionally storage-agnostic.
 */

import { TDAMError } from "./errors.js";
import { createHmac, createHash } from "node:crypto";

// ============================
// Platform response types
// ============================

/** Raw response from platform `POST /v2/cos/secret`. */
export interface CosSecretResponse {
  CosUrl: string;
  TmpSecretId: string;
  TmpSecretKey: string;
  TmpToken: string;
  ExpirationTime: string;
  PathPrefix: string;
}

// ============================
// COS URL parser
// ============================

/**
 * Parse CosUrl → { bucket, region, host }.
 *
 * Supports:
 *   - Public:   https://bucket.cos.region.myqcloud.com
 *   - Internal: https://bucket.cos-internal.region.tencentcos.cn
 */
function parseCosUrl(cosUrl: string): { bucket: string; region: string; host: string } {
  let host: string;
  try {
    host = new URL(cosUrl).hostname;
  } catch {
    throw new TDAMError(-1, `Invalid CosUrl: ${cosUrl}`);
  }
  // Public: {bucket}.cos.{region}.myqcloud.com
  const pub = host.match(/^(.+?)\.cos\.(.+?)\.myqcloud\.com$/);
  if (pub) return { bucket: pub[1]!, region: pub[2]!, host };
  // Internal: {bucket}.cos-internal.{region}.tencentcos.cn
  const priv = host.match(/^(.+?)\.cos-internal\.(.+?)\.tencentcos\.cn$/);
  if (priv) return { bucket: priv[1]!, region: priv[2]!, host };
  throw new TDAMError(-1, `Cannot parse CosUrl: ${cosUrl}`);
}

// ============================
// STS Credential
// ============================

export class StsCredential {
  readonly tmpSecretId: string;
  readonly tmpSecretKey: string;
  readonly token: string;
  readonly bucket: string;
  readonly region: string;
  readonly prefix: string;
  readonly expiresAtMs: number;
  private readonly _host: string;

  constructor(data: CosSecretResponse) {
    this.tmpSecretId = data.TmpSecretId;
    this.tmpSecretKey = data.TmpSecretKey;
    this.token = data.TmpToken || "";
    const parsed = parseCosUrl(data.CosUrl);
    this.bucket = parsed.bucket;
    this.region = parsed.region;
    this._host = parsed.host;
    // Ensure trailing slash for key concatenation
    const pfx = data.PathPrefix || "";
    this.prefix = pfx.endsWith("/") ? pfx : `${pfx}/`;
    this.expiresAtMs = data.ExpirationTime
      ? new Date(data.ExpirationTime).getTime()
      : Date.now() + 30 * 60 * 1000;
  }

  isValid(bufferMs = 120_000): boolean {
    return Date.now() < this.expiresAtMs - bufferMs;
  }

  get cosHost(): string {
    return this._host;
  }
}

// ============================
// STS Credential Manager
// ============================

export interface MemoryFileReaderConfig {
  endpoint: string;
  apiKey: string;
  serviceId: string;
  timeout?: number;
}

export class StsCredentialManager {
  private credential: StsCredential | null = null;
  private fetchPromise: Promise<StsCredential> | null = null;
  private readonly bufferMs: number;
  private readonly endpoint: string;
  private readonly apiKey: string;
  private readonly serviceId: string;
  private readonly timeout: number;

  constructor(
    config: MemoryFileReaderConfig,
    bufferMs = 120_000,
  ) {
    this.endpoint = config.endpoint.replace(/\/+$/, "");
    this.apiKey = config.apiKey;
    this.serviceId = config.serviceId;
    this.timeout = config.timeout ?? 30_000;
    this.bufferMs = bufferMs;
  }

  async getCredential(): Promise<StsCredential> {
    if (this.credential?.isValid(this.bufferMs)) {
      return this.credential;
    }
    // Coalesce concurrent requests
    if (!this.fetchPromise) {
      this.fetchPromise = this.refresh();
    }
    try {
      return await this.fetchPromise;
    } finally {
      this.fetchPromise = null;
    }
  }

  invalidate(): void {
    this.credential = null;
    this.fetchPromise = null;
  }

  private async refresh(): Promise<StsCredential> {
    const url = `${this.endpoint}/v2/cos/secret`;
    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), this.timeout);

    try {
      const resp = await fetch(url, {
        method: "POST",
        headers: {
          Authorization: `Bearer ${this.apiKey}`,
          "x-tdai-service-id": this.serviceId,
          "Content-Type": "application/json",
        },
        body: "{}",
        signal: controller.signal,
      });

      if (!resp.ok) {
        const text = await resp.text().catch(() => "");
        throw new TDAMError(resp.status, `COS secret fetch failed: HTTP ${resp.status}: ${text}`);
      }

      const data = (await resp.json()) as CosSecretResponse;
      const cred = new StsCredential(data);
      this.credential = cred;
      return cred;
    } finally {
      clearTimeout(timer);
    }
  }
}

// ============================
// COS V5 Signature (GET only)
// ============================

/**
 * Generate COS V5 Authorization header for a GET request.
 * Uses Node.js crypto.
 */
export function cosV5Sign(
  secretId: string,
  secretKey: string,
  method: string,
  path: string,
  host: string,
  startTime?: number,
  endTime?: number,
): string {
  const now = Math.floor(Date.now() / 1000);
  const qSignTime = `${startTime ?? now - 60};${endTime ?? now + 600}`;
  const qKeyTime = qSignTime;

  // Step 1: SignKey = HMAC-SHA1(SecretKey, q-key-time)
  const signKey = hmacSha1Hex(secretKey, qKeyTime);

  // Step 2: HttpString
  const httpString = `${method.toLowerCase()}\n${path}\n\nhost=${host}\n`;

  // Step 3: StringToSign
  const sha1HttpString = sha1Hex(httpString);
  const stringToSign = `sha1\n${qSignTime}\n${sha1HttpString}\n`;

  // Step 4: Signature
  const signature = hmacSha1Hex(signKey, stringToSign);

  return (
    `q-sign-algorithm=sha1` +
    `&q-ak=${secretId}` +
    `&q-sign-time=${qSignTime}` +
    `&q-key-time=${qKeyTime}` +
    `&q-header-list=host` +
    `&q-url-param-list=` +
    `&q-signature=${signature}`
  );
}

// Node.js crypto helpers
function hmacSha1Hex(key: string, data: string): string {
  return createHmac("sha1", key).update(data).digest("hex");
}

function sha1Hex(data: string): string {
  return createHash("sha1").update(data).digest("hex");
}

// ============================
// Memory File Reader
// ============================

export class MemoryFileReader {
  constructor(
    private readonly stsManager: StsCredentialManager,
    private readonly timeout = 30_000,
  ) {}

  /**
   * Read a memory file (persona.md, scene_blocks/*.md, …) by relative path.
   *
   * @param path Relative path within the memory space,
   *   e.g. "scene_blocks/cooking-recipes.md" or "persona.md".
   * @returns File content as UTF-8 string.
   */
  async read(path: string): Promise<string> {
    let cred = await this.stsManager.getCredential();
    let result = await this.doGet(cred, path);

    // 403 → invalidate and retry once
    if (result.status === 403) {
      this.stsManager.invalidate();
      cred = await this.stsManager.getCredential();
      result = await this.doGet(cred, path);
    }

    if (result.status === 404) {
      throw new TDAMError(404, `File not found: ${path}`);
    }

    if (result.status !== 200) {
      throw new TDAMError(
        result.status,
        `COS GET failed: HTTP ${result.status} — ${result.body.slice(0, 200)}`,
      );
    }

    return result.body;
  }

  private async doGet(
    cred: StsCredential,
    path: string,
  ): Promise<{ status: number; body: string }> {
    const fullKey = `${cred.prefix}${path}`;
    const cosPath = `/${fullKey}`;
    const host = cred.cosHost;

    const auth = cosV5Sign(
      cred.tmpSecretId,
      cred.tmpSecretKey,
      "GET",
      cosPath,
      host,
    );

    const headers: Record<string, string> = {
      Host: host,
      Authorization: auth,
    };
    if (cred.token) {
      headers["x-cos-security-token"] = cred.token;
    }

    const url = `https://${host}${cosPath}`;
    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), this.timeout);

    try {
      const resp = await fetch(url, {
        method: "GET",
        headers,
        signal: controller.signal,
      });
      const body = await resp.text();
      return { status: resp.status, body };
    } finally {
      clearTimeout(timer);
    }
  }
}

/**
 * Convenience factory: create a MemoryFileReader that fetches STS credentials
 * directly from the platform's `POST /v2/cos/secret` endpoint.
 */
export function createMemoryFileReader(
  config: MemoryFileReaderConfig,
  opts?: { bufferMs?: number; timeout?: number },
): MemoryFileReader {
  const mgr = new StsCredentialManager(config, opts?.bufferMs);
  return new MemoryFileReader(mgr, opts?.timeout);
}

File: sdk/memory-core/python/tencentdb_agent_memory/cos.py
"""Memory file reader — direct read of memory pipeline artifacts (persona.md,
scene_blocks/*.md) from object storage with STS credential management.

The SDK user calls ``client.read_file(path)`` and gets the file content back.
Under the hood, we:
  1. Fetch STS temporary credentials from the platform (``POST /v2/cos/secret``)
  2. Cache credentials until they expire (auto-refresh)
  3. Sign a COS V5 GET request with the STS credentials
  4. Return the file content as a string

Storage backend (currently COS) is an implementation detail — the public API
is intentionally storage-agnostic.
"""

from __future__ import annotations

import hashlib
import hmac
import logging
import re
import threading
import time
from typing import Any, Dict, Optional
from urllib.parse import urlparse

import httpx

from .errors import TDAMError

logger = logging.getLogger(__name__)


# ---------------------------------------------------------------------------
# COS URL parser
# ---------------------------------------------------------------------------

def _parse_cos_url(cos_url: str) -> tuple[str, str]:
    """Parse CosUrl like ``https://bucket.cos.region.myqcloud.com`` → (bucket, region)."""
    host = urlparse(cos_url).hostname or ""
    # Pattern: {bucket}.cos.{region}.myqcloud.com
    m = re.match(r"^(.+?)\.cos\.(.+?)\.myqcloud\.com$", host)
    if m:
        return m.group(1), m.group(2)
    raise TDAMError(
        code=-1,
        message=f"Cannot parse CosUrl: {cos_url!r} (expected {{bucket}}.cos.{{region}}.myqcloud.com)",
    )


# ---------------------------------------------------------------------------
# STS Credential
# ---------------------------------------------------------------------------

class StsCredential:
    """Parsed STS credential from ``POST /v2/cos/secret``.

    Platform response format::

        {
            "CosUrl": "https://{bucket}.cos.{region}.myqcloud.com",
            "TmpSecretId": "...",
            "TmpSecretKey": "...",
            "TmpToken": "...",
            "ExpirationTime": "2026-05-15T16:44:49+08:00",
            "PathPrefix": "memory_v2/cos_data/mem-xxx"
        }
    """

    __slots__ = (
        "tmp_secret_id", "tmp_secret_key", "token",
        "bucket", "region", "prefix", "expires_at_epoch",
    )

    def __init__(self, data: Dict[str, Any]) -> None:
        self.tmp_secret_id: str = data["TmpSecretId"]
        self.tmp_secret_key: str = data["TmpSecretKey"]
        self.token: str = data.get("TmpToken", "")
        # Parse bucket + region from CosUrl
        self.bucket: str
        self.region: str
        self.bucket, self.region = _parse_cos_url(data["CosUrl"])
        # PathPrefix — ensure trailing slash for key concatenation
        prefix = data.get("PathPrefix", "")
        self.prefix: str = prefix if prefix.endswith("/") else f"{prefix}/"
        # Parse ISO 8601 → epoch seconds
        expires_str = data.get("ExpirationTime", "")
        if expires_str:
            from datetime import datetime, timezone
            try:
                dt = datetime.fromisoformat(expires_str.replace("Z", "+00:00"))
                self.expires_at_epoch = dt.timestamp()
            except Exception:
                self.expires_at_epoch = time.time() + 1800  # 30 min fallback
        else:
            self.expires_at_epoch = time.time() + 1800

    def is_valid(self, buffer_seconds: float = 120) -> bool:
        """Check if credential is still valid (with 2-minute buffer)."""
        return time.time() < (self.expires_at_epoch - buffer_seconds)

    @property
    def cos_host(self) -> str:
        return f"{self.bucket}.cos.{self.region}.myqcloud.com"


# ---------------------------------------------------------------------------
# STS Credential Manager
# ---------------------------------------------------------------------------

class StsCredentialManager:
    """Thread-safe STS credential cache with auto-refresh.

    - Fetches STS from platform ``POST /v2/cos/secret``
    - Caches until expiry (with 2-minute buffer)
    - Coalesces concurrent refresh requests
    """

    def __init__(
        self,
        endpoint: str,
        api_key: str,
        service_id: str,
        buffer_seconds: float = 120,
        timeout: float = 30,
    ) -> None:
        self._endpoint = endpoint.rstrip("/")
        self._api_key = api_key
        self._service_id = service_id
        self._buffer = buffer_seconds
        self._timeout = timeout
        self._credential: Optional[StsCredential] = None
        self._lock = threading.Lock()
        self._client: Optional[httpx.Client] = None

    def get_credential(self) -> StsCredential:
        """Get a valid STS credential (cached or freshly fetched)."""
        if self._credential and self._credential.is_valid(self._buffer):
            return self._credential

        with self._lock:
            # Double-check after acquiring lock
            if self._credential and self._credential.is_valid(self._buffer):
                return self._credential
            return self._refresh()

    def invalidate(self) -> None:
        """Force invalidate cached credential (e.g. on 403)."""
        with self._lock:
            self._credential = None

    def _refresh(self) -> StsCredential:
        logger.debug("[cos] Refreshing STS credential via POST /v2/cos/secret ...")
        if self._client is None:
            self._client = httpx.Client(timeout=self._timeout)

        url = f"{self._endpoint}/v2/cos/secret"
        headers = {
            "Authorization": f"Bearer {self._api_key}",
            "x-tdai-service-id": self._service_id,
            "Content-Type": "application/json",
        }
        resp = self._client.post(url, json={}, headers=headers)
        resp.raise_for_status()
        data = resp.json()

        cred = StsCredential(data)
        self._credential = cred
        logger.debug("[cos] STS refreshed: bucket=%s prefix=%s expires=%.0f",
                     cred.bucket, cred.prefix, cred.expires_at_epoch)
        return cred

    def close(self) -> None:
        if self._client is not None:
            self._client.close()


# ---------------------------------------------------------------------------
# Async STS Credential Manager
# ---------------------------------------------------------------------------

class AsyncStsCredentialManager:
    """Async variant of StsCredentialManager."""

    def __init__(
        self,
        endpoint: str,
        api_key: str,
        service_id: str,
        buffer_seconds: float = 120,
        timeout: float = 30,
    ) -> None:
        self._endpoint = endpoint.rstrip("/")
        self._api_key = api_key
        self._service_id = service_id
        self._buffer = buffer_seconds
        self._timeout = timeout
        self._credential: Optional[StsCredential] = None
        import asyncio
        self._lock = asyncio.Lock()
        self._client: Optional[httpx.AsyncClient] = None

    async def get_credential(self) -> StsCredential:
        if self._credential and self._credential.is_valid(self._buffer):
            return self._credential

        async with self._lock:
            if self._credential and self._credential.is_valid(self._buffer):
                return self._credential
            return await self._refresh()

    def invalidate(self) -> None:
        self._credential = None

    async def _refresh(self) -> StsCredential:
        logger.debug("[cos] Refreshing STS credential (async) via POST /v2/cos/secret ...")
        if self._client is None:
            self._client = httpx.AsyncClient(timeout=self._timeout)

        url = f"{self._endpoint}/v2/cos/secret"
        headers = {
            "Authorization": f"Bearer {self._api_key}",
            "x-tdai-service-id": self._service_id,
            "Content-Type": "application/json",
        }
        resp = await self._client.post(url, json={}, headers=headers)
        resp.raise_for_status()
        data = resp.json()

        cred = StsCredential(data)
        self._credential = cred
        return cred

    async def close(self) -> None:
        if self._client is not None:
            await self._client.aclose()


# ---------------------------------------------------------------------------
# COS V5 Signature
# ---------------------------------------------------------------------------

def _cos_v5_sign(
    secret_id: str,
    secret_key: str,
    method: str,
    path: str,
    host: str,
    start_time: Optional[int] = None,
    end_time: Optional[int] = None,
) -> str:
    """Generate COS V5 Authorization header value for a GET request.

    References:
      https://cloud.tencent.com/document/product/436/7778
    """
    now = int(time.time())
    q_sign_time = f"{start_time or (now - 60)};{end_time or (now + 600)}"
    q_key_time = q_sign_time

    # Step 1: SignKey
    sign_key = hmac.new(
        secret_key.encode("utf-8"),
        q_key_time.encode("utf-8"),
        hashlib.sha1,
    ).hexdigest()

    # Step 2: HttpString
    # For GET with no query params and only host header
    http_string = f"{method.lower()}\n{path}\n\nhost={host}\n"

    # Step 3: StringToSign
    sha1_http_string = hashlib.sha1(http_string.encode("utf-8")).hexdigest()
    string_to_sign = f"sha1\n{q_sign_time}\n{sha1_http_string}\n"

    # Step 4: Signature
    signature = hmac.new(
        sign_key.encode("utf-8"),
        string_to_sign.encode("utf-8"),
        hashlib.sha1,
    ).hexdigest()

    return (
        f"q-sign-algorithm=sha1"
        f"&q-ak={secret_id}"
        f"&q-sign-time={q_sign_time}"
        f"&q-key-time={q_key_time}"
        f"&q-header-list=host"
        f"&q-url-param-list="
        f"&q-signature={signature}"
    )


# ---------------------------------------------------------------------------
# Memory File Reader (sync)
# ---------------------------------------------------------------------------

class MemoryFileReader:
    """Sync memory file reader with STS auto-management.

    Reads memory pipeline artifacts (persona.md, scene_blocks/*.md, …) from
    object storage. The storage backend is COS today but the public API is
    storage-agnostic.

    Usage::

        reader = MemoryFileReader(sts_manager)
        content = reader.read("scene_blocks/cooking-recipes.md")
    """

    def __init__(
        self,
        sts_manager: StsCredentialManager,
        timeout: float = 30,
        client: Optional[httpx.Client] = None,
    ) -> None:
        self._sts = sts_manager
        self._client = client or httpx.Client(timeout=timeout)

    def read(self, path: str) -> str:
        """Read a memory file by relative path.

        The final COS key is: ``{prefix}{path}``

        Returns file content as UTF-8 string.
        Raises ``TDAMError`` on failure.
        """
        cred = self._sts.get_credential()
        full_key = f"{cred.prefix}{path}"
        cos_path = f"/{full_key}"
        host = cred.cos_host

        auth = _cos_v5_sign(
            secret_id=cred.tmp_secret_id,
            secret_key=cred.tmp_secret_key,
            method="GET",
            path=cos_path,
            host=host,
        )

        headers: Dict[str, str] = {
            "Host": host,
            "Authorization": auth,
        }
        if cred.token:
            headers["x-cos-security-token"] = cred.token

        url = f"https://{host}{cos_path}"
        logger.debug("[cos] GET %s", url)

        resp = self._client.get(url, headers=headers)

        if resp.status_code == 403:
            # Invalidate and retry once
            logger.warning("[cos] 403 on GET %s — invalidating STS and retrying", path)
            self._sts.invalidate()
            cred = self._sts.get_credential()
            full_key = f"{cred.prefix}{path}"
            cos_path = f"/{full_key}"
            host = cred.cos_host
            auth = _cos_v5_sign(
                secret_id=cred.tmp_secret_id,
                secret_key=cred.tmp_secret_key,
                method="GET",
                path=cos_path,
                host=host,
            )
            headers = {"Host": host, "Authorization": auth}
            if cred.token:
                headers["x-cos-security-token"] = cred.token
            url = f"https://{host}{cos_path}"
            resp = self._client.get(url, headers=headers)

        if resp.status_code == 404:
            raise TDAMError(code=404, message=f"File not found: {path}")

        if resp.status_code != 200:
            raise TDAMError(
                code=resp.status_code,
                message=f"COS GET failed: HTTP {resp.status_code} — {resp.text[:200]}",
            )

        return resp.text

    def close(self) -> None:
        if isinstance(self._client, httpx.Client):
            self._client.close()


# ---------------------------------------------------------------------------
# Async Memory File Reader
# ---------------------------------------------------------------------------

class AsyncMemoryFileReader:
    """Async memory file reader with STS auto-management."""

    def __init__(
        self,
        sts_manager: AsyncStsCredentialManager,
        timeout: float = 30,
        client: Optional[httpx.AsyncClient] = None,
    ) -> None:
        self._sts = sts_manager
        self._client = client or httpx.AsyncClient(timeout=timeout)

    async def read(self, path: str) -> str:
        cred = await self._sts.get_credential()
        full_key = f"{cred.prefix}{path}"
        cos_path = f"/{full_key}"
        host = cred.cos_host

        auth = _cos_v5_sign(
            secret_id=cred.tmp_secret_id,
            secret_key=cred.tmp_secret_key,
            method="GET",
            path=cos_path,
            host=host,
        )

        headers: Dict[str, str] = {
            "Host": host,
            "Authorization": auth,
        }
        if cred.token:
            headers["x-cos-security-token"] = cred.token

        url = f"https://{host}{cos_path}"
        resp = await self._client.get(url, headers=headers)

        if resp.status_code == 403:
            self._sts.invalidate()
            cred = await self._sts.get_credential()
            full_key = f"{cred.prefix}{path}"
            cos_path = f"/{full_key}"
            host = cred.cos_host
            auth = _cos_v5_sign(
                secret_id=cred.tmp_secret_id,
                secret_key=cred.tmp_secret_key,
                method="GET",
                path=cos_path,
                host=host,
            )
            headers = {"Host": host, "Authorization": auth}
            if cred.token:
                headers["x-cos-security-token"] = cred.token
            url = f"https://{host}{cos_path}"
            resp = await self._client.get(url, headers=headers)

        if resp.status_code == 404:
            raise TDAMError(code=404, message=f"File not found: {path}")

        if resp.status_code != 200:
            raise TDAMError(
                code=resp.status_code,
                message=f"COS GET failed: HTTP {resp.status_code} — {resp.text[:200]}",
            )

        return resp.text

    async def close(self) -> None:
        if isinstance(self._client, httpx.AsyncClient):
            await self._client.aclose()
